metapool/sample_sheet.py

```diff
@@ -377,73 +377,62 @@ def merge(self, sheets):
-            out = table.copy(deep=True)
-
             # if a column named 'index' is present in table, assume it is a
             # numeric index and not a sequence of bases, which is required in
             # the output. Assume the column that will become 'index' is
             # defined in remapper.
-            if 'index' in set(out.columns):
-                out.drop(columns=['index'], inplace=True)
-
-            # if an alternate form of a column name defined in
-            # _KL_SAMPLE_SHEET_COLUMN_ALTS is found in table, assume it should
-            # be renamed to its proper form and be included in the output e.g.:
-            # 'sample_plate' -> 'Sample_Plate'.
+            if 'index' in set(result.columns):
+                result.drop(columns=['index'], inplace=True)
 
-            # assume keys in _KL_SAMPLE_SHEET_COLUMN_ALTS do not overlap w/
-            # remapper (they currently do not). Define the full set of
-            # potential columns to rename in table.
-
-            # new syntax in 3.9 allows us to merge two dicts together w/OR.
             remapper = KLSampleSheet.column_alts | self.remapper
-            out.rename(remapper, axis=1, inplace=True)
+            result.rename(remapper, axis=1, inplace=True)
 
-            # out may contain additional columns that aren't allowed in the
+            # result may contain additional columns that aren't allowed in the
             # [Data] section of a sample-sheet e.g.: 'Extraction Kit Lot'.
-            # There may also be required columns that aren't defined in out.
-            subset = list(
-                set(self.data_columns) & set(
-                    out.columns))
+            # There may also be required columns that aren't defined in result.
 
-            out = out[subset]
+            # once all columns have been renamed to their preferred names, we
+            # must determine the proper set of column names for this sample-
+            # sheet. For legacy classes this is simply the list of columns
+            # defined in each sample-sheet version. For newer classes, this is
+            # defined at run-time and requires examining the metadata that
+            # will define the [Data] section.
+            result.to_csv('result.csv')
+            required_columns = self._get_expected_columns(table=result)
+            subset = list(set(required_columns) & set(result.columns))
+            result = result[subset]
 
-        # append the new 'Well_description' column, now that alternates have
-        # been removed and non-essential columns have been dropped.
-        out['Well_description'] = well_description
+        return result
 
-        for column in self.data_columns:
-            if column not in out.columns:
-                warnings.warn('The column %s in the sample sheet is empty' %
-                              column)
-                out[column] = ''
+    def _add_data_to_sheet(self, table, sequencer, lanes, assay, strict=True):
```

Comment:
Just to confirm, this change is so it becomes part of the object, right? If yes, why does it need to be named with 
and initial `_`

Reply:
_remap_table() became a member function when we implemented different sample-sheet versions as child classes. However, both before and after, it's just a helper method that is only called in one location, by _add_data_to_sheet().

At different times we added a forced overwrite of well_description column to plate.sample.well_id and we added this warning to resolve other issues. It was better to migrate out those 'bonus functionalities' out of _remap_table() and into the sole caller of the method. It clarifies things with unittests as well.